compiler/rustc_errors/src/diagnostic.rs

```diff
@@ -107,6 +108,31 @@ pub struct Diagnostic {
     /// If diagnostic is from Lint, custom hash function ignores notes
     /// otherwise hash is based on the all the fields
     pub is_lint: bool,
+
+    /// With `-Ztrack_diagnostics` enabled,
+    /// we print where in rustc this error was emitted.
+    pub emitted_at: DiagnosticLocation,
+}
+
+#[derive(Clone, Debug, Encodable, Decodable)]
+pub struct DiagnosticLocation {
+    file: Cow<'static, str>,
```

Comment:
Is there a reason to use `Cow`? From what I can see this is only initialized with `Location::caller().file().into()`, so this is always `Borrowed(&'static str)`, or am I missing something?

Moreover, can't this just be a wrapper around `panic::Location`? Or maybe even just use `Location` without this wrapper?

Reply:
These are needed because of the `Encodable` and `Decodable` derives on `Diagnostic`  because it owns a `DiagnosticLocation`. Perhaps there's a trick that I'm not aware of, but this is the simplest way I could get that to work. I don't see how I could implement these traits for `&'static str` or for `Location`.

Reply:
I see, thanks.